From 27e0dac05b05b2fe780bbcb685933a714be10ef0 Mon Sep 17 00:00:00 2001
From: Evan Huus <eapache@gmail.com>
Date: Sun, 19 Apr 2015 20:19:54 -0400
Subject: [PATCH 05/10] reassembly: address a 15-year old XXX comment

Question: "what if we didn't capture the entire fragment due to a too-short
          snapshot length?"
Answer: An assertion fails and we leak a bunch of memory.

Don't do that.

Bug: 11129
Change-Id: I0adfb217f0e66ae8f5f6255a4caf7ff940826b59
Reviewed-on: https://code.wireshark.org/review/8128
Petri-Dish: Evan Huus <eapache@gmail.com>
Tested-by: Petri Dish Buildbot <buildbot-no-reply@wireshark.org>
Reviewed-by: Evan Huus <eapache@gmail.com>
(cherry picked from commit c35f2ccb4433718416551cc7a85afb0860529d57)
Reviewed-on: https://code.wireshark.org/review/8149
(cherry picked from commit 5bbf0d2afbf31f50eed0d46a56a2a1e8f494901a)
Reviewed-on: https://code.wireshark.org/review/8638
Reviewed-by: Balint Reczey <balint@balintreczey.hu>
---
 epan/reassemble.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/epan/reassemble.c b/epan/reassemble.c
index 87b8e9b..1a6627d 100644
--- a/epan/reassemble.c
+++ b/epan/reassemble.c
@@ -1013,9 +1013,11 @@ fragment_add_work(fragment_head *fd_head, tvbuff_t *tvb, const int offset,
 
 	/* If we have reached this point, the packet is not defragmented yet.
 	 * Save all payload in a buffer until we can defragment.
-	 * XXX - what if we didn't capture the entire fragment due
-	 * to a too-short snapshot length?
 	 */
+	if (!tvb_bytes_exist(tvb, offset, fd->len)) {
+		g_slice_free(fragment_item, fd);
+		THROW(BoundsError);
+	}
 	fd->tvb_data = tvb_clone_offset_len(tvb, offset, fd->len);
 	LINK_FRAG(fd_head,fd);
 
-- 
2.1.4

